Kafka microservices update (#1260)
oktadeveloper/okta-blog: _source/_posts/2022-09-19-kafka-microservices.md

- "Did you know @jhipster supports #Kafka as a communications mechanism between microservices?"
- "Microservice architectures ❤️ asynchronous messaging! Learn how to use @apachekafka to implement it. →"
image: blog/kafka-microservices/kafka-microservices.png
github: https://github.com/oktadev/okta-kafka-microservices-example
type: conversion
---

One of the traditional approaches for communicating between [microservices](https://www.okta.com/blog/2021/02/microservices/) is through their REST APIs. However, as your system evolves and the number of microservices grows, communication becomes more complex, and the architecture might start resembling our old friend the spaghetti anti-pattern, with services depending on each other or tightly coupled, slowing down development teams. This model can exhibit low latency but only works if services are made highly available.

To overcome this design disadvantage, new architectures aim to decouple senders from receivers, with asynchronous messaging. In a Kafka-centric architecture, low latency is preserved, with additional advantages like message balancing among available consumers and centralized management.

> When dealing with a brownfield platform (legacy), a recommended way to de-couple a monolith and ready it for a move to microservices is to implement asynchronous messaging.

Reviewer: ```suggestion
When dealing with a brownfield platform (legacy), a recommended way to decouple a monolith and ready it for a move to microservices is to implement asynchronous messaging.
```
decouple is one word

In this tutorial you will learn how to:

- Create a microservices architecture with JHipster
- Enable Kafka integration for communicating microservices
- Set up Okta as the authentication provider

{% include toc.md %}

## What is Kafka?

**Apache Kafka** is a distributed streaming platform. It was initially conceived as a message queue and open-sourced by LinkedIn in 2011. Its community evolved Kafka to provide key capabilities:

- **Publish and Subscribe** to streams of records, like a message queue.
- **Storage system** so messages can be consumed asynchronously. Kafka writes data to a scalable disk structure and replicates it for fault-tolerance. Producers can wait for write acknowledgments.

> - **Stream processing** with Kafka Streams API, enables complex aggregations or joins of input streams onto an output stream of processed data.

Reviewer: ```suggestion
- **Stream processing**, with Kafka Streams API, enables complex aggregations or joins of input streams onto an output stream of processed data.
```

Traditional messaging models are queue and publish-subscribe. In a queue, each record goes to one consumer. In publish-subscribe, the record is received by all consumers.

> The **Consumer Group** in Kafka is an abstraction that combines both models. Record processing can be load balanced among the members of a consumer group and Kafka allows to broadcast messages to multiple consumer groups. It is the same publish-subscribe semantic where the subscriber is a cluster of consumers instead of a single process.

Reviewer: ```suggestion
The **Consumer Group** in Kafka is an abstraction that combines both models. Record processing can be load balanced among the members of a consumer group and Kafka allows you to broadcast messages to multiple consumer groups. It is the same publish-subscribe semantic where the subscriber is a cluster of consumers instead of a single process.
```
allow requires an object

Popular use cases of Kafka include:

> - The traditional messaging, to decouple data producers from processors with better latency and scalability.

Reviewer: ```suggestion
- Traditional messaging, to decouple data producers from processors with better latency and scalability.
```
consistency of the bullet formats, and consistent punctuation.

> - Site activity tracking with real-time publish-subscribe feeds

Reviewer: ```suggestion
- Site activity tracking with real-time publish-subscribe feeds.
```

> - As a replacement for file-based log aggregation, where event data becomes a stream of messages

Reviewer: ```suggestion
- As a replacement for file-based log aggregation, where event data becomes a stream of messages.
```
period for visual consistency in the bulleted list.

> - Data Pipelines where data consumed from topics is transformed and fed to new topics

Reviewer: ```suggestion
- Data pipelines where data consumed from topics is transformed and fed to new topics.
```

> - As an external commit log for a distributed system

Reviewer: ```suggestion
- As an external commit log for a distributed system.
```

> - As a backend log storage for event sourcing applications, where each state change is logged in time order.

Reviewer: ```suggestion
- As backend log storage for event sourcing applications, where each state change is logged in time order.
```

> ## Microservices Communication With Kafka

Reviewer: ```suggestion
## Microservices communication With Kafka
```
sentence case

> Let's build a microservices architecture with JHipster and Kafka support. In this tutorial, you'll create a `store` and an `alert` microservices. The store microservices will create and update store records. The `alert` microservice will receive update events from `store` and send an email alert.

Reviewer: ```suggestion
Let's build a microservices architecture with JHipster and Kafka support. In this tutorial, you'll create `store` and `alert` microservices. The store microservices will create and update store records. The `alert` microservice will receive update events from `store` and send an email alert.
```

    testFrameworks [cypress]
  }
  entities Store, StoreAlert
}

application {
  config {
    baseName store,
    packageName com.okta.developer.store,
    applicationType microservice,
    authenticationType oauth2,
    databaseType mongodb,
    devDatabaseType mongodb,
    prodDatabaseType mongodb,
    enableHibernateCache false,
    serverPort 8082,
    serviceDiscoveryType eureka
    messageBroker kafka
  }
  entities Store
}

application {
  config {
    baseName alert,
    packageName com.okta.developer.alert,
    applicationType microservice,
    authenticationType oauth2,
    serverPort 8082,
    serviceDiscoveryType eureka
    messageBroker kafka
  }
  entities StoreAlert
}

enum StoreStatus {
  OPEN,
  CLOSED
}

entity Store {
  name String required,
  address String required,
  status StoreStatus,
  createTimestamp Instant required,
  updateTimestamp Instant
}

entity StoreAlert {
  storeName String required,
  storeStatus String required,
  timestamp Instant required
}

microservice Store with store
microservice StoreAlert with alert
```

Now, in your `jhipster-kafka` folder, import this file with the following command:

```bash
jhipster jdl apps.jdl
```

> ## Configure Microservices Deployment with Docker Compose

Reviewer: ```suggestion
## Configure microservices deployment with Docker Compose
```

cd docker-compose
jhipster docker-compose
```

The generator will ask you to define the following things:

1. Type of application: **Microservice application**
2. Type of gateway: **JHipster gateway based on Spring Cloud Gateway**
3. Leave the root directory for services as default: **../**
4. Which applications to include: **gateway**, **store**, **alert**
4. Which applications do you want to use with clustered databases: **(none)**
5. If monitoring should be enabled: **No**
6. Password for JHipster Registry: `<default>`

Almost when the generator completes, a warning shows in the output:

```bash
WARNING! Docker Compose configuration generated, but no Jib cache found
If you forgot to generate the Docker image for this application, please run:
To generate the missing Docker image(s), please run:
 ./mvnw -ntp -Pprod verify jib:dockerBuild in /home/indiepopart/jhipster-kafka/alert
 ./mvnw -ntp -Pprod verify jib:dockerBuild in /home/indiepopart/jhipster-kafka/gateway
 ./mvnw -ntp -Pprod verify jib:dockerBuild in /home/indiepopart/jhipster-kafka/store
```

You will generate the images later, but first, let's add some security and Kafka integration to your microservices.

> ## Add OpenID Connect (OIDC) Authentication

Reviewer: ```suggestion
## Add OpenID Connect (OIDC) authentication
```

This microservices architecture is set up to authenticate against Keycloak. Let's update the settings to use Okta as the authentication provider.

{% include setup/cli.md type="jhipster" loginRedirectUri="http://localhost:8081/login/oauth2/code/oidc,http://localhost:8761/login/oauth2/code/oidc" logoutRedirectUri="http://localhost:8081,http://localhost:8761" %}

In the project, create a `docker-compose/.env` file and add the following variables. For the values, use the settings from the Okta web application you created:

```bash
OIDC_ISSUER_URI={yourIssuer}
OIDC_CLIENT_ID={yourClientId}
OIDC_CLIENT_SECRET={yourClientSecret}
```

Edit `docker-compose/docker-compose.yml` and update the `SPRING_SECURITY_*` settings for the services `store-app`, `alert-app`, `gateway-app`, and `jhipster-registry`:

```yaml
- SPRING_SECURITY_OAUTH2_CLIENT_PROVIDER_OIDC_ISSUER_URI=${OIDC_ISSUER_URI}
- SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_OIDC_CLIENT_ID=${OIDC_CLIENT_ID}
- SPRING_SECURITY_OAUTH2_CLIENT_REGISTRATION_OIDC_CLIENT_SECRET=${OIDC_CLIENT_SECRET}
```

> ### Use Spring Cloud Config to Override OpenID Connect Settings

Reviewer: ```suggestion
### Use Spring Cloud Config to override OpenID Connect settings
```

An alternative to setting environment variables for each application in `docker-compose.yml` is to use Spring Cloud Config. JHipster Registry includes Spring Cloud Config, so it's pretty easy to do.

Open `docker-compose/central-server-config/application.yml` and add your Okta settings there.

```yaml
spring:
  security:
    oauth2:
      client:
        provider:
          oidc:
            issuer-uri: https://{yourOktaDomain}/oauth2/default
        registration:
          oidc:
            client-id: {yourClientId}
            client-secret: {yourClientSecret}
```

The registry, gateway, store, and alert applications are all configured to read this configuration on startup.

> ## Communicate Between Store and Alert Microservices

Reviewer: ```suggestion
## Communicate between store and alert microservices
```

    }

    public String getStoreStatus() {
        return storeStatus;
    }

    public void setStoreStatus(String storeStatus) {
        this.storeStatus = storeStatus;
    }

}
```

Inject the `AlertService` into the `StoreResource` API implementation, modifying its constructor. Also modify the `updateStore` call to publish a `StoreAlertDTO` for the `alert` service:

```java
@RestController
@RequestMapping("/api")
public class StoreResource {

    ...
    private final StoreRepository storeRepository;
    private final AlertService alertService;

    public StoreResource(StoreRepository storeRepository, AlertService alertService) {
        this.storeRepository = storeRepository;
        this.alertService = alertService;
    }

    ...

    @PutMapping("/stores/{id}")
    public ResponseEntity<Store> updateStore(
        @PathVariable(value = "id", required = false) final String id,
        @Valid @RequestBody Store store
    ) throws URISyntaxException {
        ...

        Store result = storeRepository.save(store);

        log.debug("SEND store alert for Store: {}", store);
        alertService.alertStoreStatus(result);

        ...
    }

   ...
}
```

> ### Enable Debug Logging in Production

Reviewer: ```suggestion
### Enable debug logging in production
```

        } catch (Exception exception) {
            throw new EmailServiceException(exception);
        }
    }
}
```

Create the referenced `EmailServiceException`.

```java
package com.okta.developer.alert.service;

public class EmailServiceException extends RuntimeException {

    public EmailServiceException(Exception exception) {
        super(exception);
    }
}
```

Add a `StoreAlertDTO` class in the `...service.dto` package.

```java
package com.okta.developer.alert.service.dto;

public class StoreAlertDTO {

    private String storeName;
    private String storeStatus;

    public String getStoreName() {
        return storeName;
    }

    public void setStoreName(String storeName) {
        this.storeName = storeName;
    }

    public String getStoreStatus() {
        return storeStatus;
    }

    public void setStoreStatus(String storeStatus) {
        this.storeStatus = storeStatus;
    }

}
```

Add a new property to `alert/src/main/resources/config/application.yml` **and** to `alert/src/test/resources/config/application.yml` for the destination email of the store alert.

```yaml
alert:
  distribution-list: {distributionListAddress}
```

> **NOTE:** You'll need to set a value for the email (e.g., `list@email.com` will work) in `src/test/.../application.yml` for tests to pass. For Docker, you'll override the `{distributionListAddress}` and `{username}` + `{password}` placeholder values with environment variables below.

Reviewer: ```suggestion
**NOTE:** You'll need to set a value for the email (e.g. `list@email.com` will work) in `src/test/.../application.yml` for tests to pass. For Docker, you'll override the `{distributionListAddress}` and `{username}` + `{password}` placeholder values with environment variables below.
```
e.g. doesn't take a comma, only "i.e., " requires one.

Update `spring.mail.*` properties in `application-prod.yml` to set Gmail as the email service:

```yaml
spring:
  ...
  mail:
    host: smtp.gmail.com
    port: 587
    username: {username}
    protocol: smtp
    tls: true
    properties.mail.smtp:
      auth: true
      starttls.enable: true
```

> ### Add a Kafka Consumer to Persist Alert and Send Email

Reviewer: ```suggestion
### Add a Kafka consumer to persist alert and send email
```

> Create an `AlertConsumer` service to persist a `StoreAlert` and send the email notification when receiving an alert message through Kafka. Add `KafkaProperties`, `StoreAlertRepository` and `EmailService` as constructor arguments. Then add a `start()` method to initialize the consumer and enter the processing loop.

Reviewer: ```suggestion
Create an `AlertConsumer` service to persist a `StoreAlert` and send the email notification when receiving an alert message through Kafka. Add `KafkaProperties`, `StoreAlertRepository`, and `EmailService` as constructor arguments. Then add a `start()` method to initialize the consumer and enter the processing loop.
```
serial comma

        this.storeAlertRepository = storeAlertRepository;
        this.emailService = emailService;
    }

    @StreamListener(value = KafkaStoreAlertConsumer.CHANNELNAME, copyHeaders = "false")
    public void consume(Message<StoreAlertDTO> message) {
        StoreAlertDTO dto = message.getPayload();
        log.info("Got message from kafka stream: {} {}", dto.getStoreName(), dto.getStoreStatus());
        try {
            StoreAlert storeAlert = new StoreAlert();
            storeAlert.setStoreName(dto.getStoreName());
            storeAlert.setStoreStatus(dto.getStoreStatus());
            storeAlert.setTimestamp(Instant.now());

            storeAlertRepository.save(storeAlert);
            emailService.sendSimpleMessage(dto);
        } catch (Exception e) {
            log.error(e.getMessage(), e);
        }
    }
}
```

**NOTE**: Any unhandled exception during message processing will make the service leave the consumer group. That's why there's code above that catches `Exception`.

As a last customization step, update the logging configuration the same way you did for the `store` microservice.

> ## Microservices + Kafka Container Deployment

Reviewer: ```suggestion
## Microservices + Kafka container deployment
```

2022-09-05 15:20:44.229  INFO 1 --- [container-0-C-1] o.a.k.c.c.internals.ConsumerCoordinator  : [Consumer clientId=consumer-store-alerts-5, groupId=store-alerts] (Re-)joining group
2022-09-05 15:20:44.238  INFO 1 --- [container-0-C-1] o.a.k.c.c.internals.ConsumerCoordinator  : [Consumer clientId=consumer-store-alerts-5, groupId=store-alerts] Request joining group due to: need to re-join with the given member-id
2022-09-05 15:20:44.239  INFO 1 --- [container-0-C-1] o.a.k.c.c.internals.ConsumerCoordinator  : [Consumer clientId=consumer-store-alerts-5, groupId=store-alerts] (Re-)joining group
```

Once everything is up, go to the gateway at `http://localhost:8081` and log in. Create a store entity and then update it. The `alert` microservice should log entries when processing the received message from the `store` service.

```bash
2022-09-05 18:08:31.546  INFO 1 --- [container-0-C-1] c.o.d.alert.service.AlertConsumer        : Got message from kafka stream: Candle Shop CLOSED
```

> If you see a `MailAuthenticationException` in the `alert` microservices log, when attempting to send the notification, it might be your Gmail security configuration.

Reviewer: ```suggestion
If you see a `MailAuthenticationException` in the `alert` microservices log when attempting to send the notification, it might be your Gmail security configuration.
```
comma not needed there.

```text
alert-app_1           | org.springframework.mail.MailAuthenticationException: Authentication failed; nested exception is javax.mail.AuthenticationFailedException: 535-5.7.8 Username and Password not accepted. Learn more at
alert-app_1           | 535 5.7.8  https://support.google.com/mail/?p=BadCredentials *** - gsmtp
alert-app_1           |
alert-app_1           | 	at org.springframework.mail.javamail.JavaMailSenderImpl.doSend(JavaMailSenderImpl.java:440)
```

> To enable the login from the `alert` application, go to [https://myaccount.google.com](https://myaccount.google.com/) and then choose the **Security** tab. Turn on 2-Step Verification for your account. In the section _Signing in to Google_, choose **App passwords** and create a new [app password](https://support.google.com/accounts/answer/185833). In _Select app_ drop down set **Other (Custom name)** and type the name for this password. Click **Generate** and copy the password. Update `docker-compose/.env` and set the app password for gmail authentication.

Reviewer: ```suggestion
To enable the login from the `alert` application, go to [https://myaccount.google.com](https://myaccount.google.com/) and then choose the **Security** tab. Turn on 2-Step Verification for your account. In the section _Signing in to Google_, choose **App passwords** and create a new [app password](https://support.google.com/accounts/answer/185833). In the _Select app_ dropdown set **Other (Custom name)** and type the name for this password. Click **Generate** and copy the password. Update `docker-compose/.env` and set the app password for gmail authentication.
```

```bash
MAIL_PASSWORD={yourAppPassword}
```

**IMPORTANT**: Don't forget to delete the app password once the test is done.

Stop all the containers with CTRL+C and restart again with `docker compose up`.
Update a store again and you should receive an email with the store's status this time.

In this tutorial, authentication (of producers and consumers), authorization (of read/write operations), and encryption (of data) were not covered, as security in Kafka is optional. See [Kafka's documentation on security](https://kafka.apache.org/documentation/#security) to learn how to enable these features.

> ## Learn More About Kafka and Microservices

Reviewer: ```suggestion
## Learn more about Kafka and microservices
```

        this.storeStatus = store.getStatus().name();
    }

    public String getStoreName() {
        return storeName;
    }

    public void setStoreName(String storeName) {
        this.storeName = storeName;
    }

    public String getStoreStatus() {
        return storeStatus;
    }

    public void setStoreStatus(String storeStatus) {
        this.storeStatus = storeStatus;
    }

}
```

Inject the `AlertService` into the `StoreResource` API implementation, modifying its constructor. Also modify the `updateStore` call to publish a `StoreAlertDTO` for the `alert` service:

```java
@RestController
@RequestMapping("/api")
public class StoreResource {

    ...
    private final StoreRepository storeRepository;
    private final AlertService alertService;

    public StoreResource(StoreRepository storeRepository, AlertService alertService) {
        this.storeRepository = storeRepository;
        this.alertService = alertService;
    }

    ...

    @PutMapping("/stores/{id}")
    public ResponseEntity<Store> updateStore(
        @PathVariable(value = "id", required = false) final String id,
        @Valid @RequestBody Store store
    ) throws URISyntaxException {

>         log.debug("REST request to update Store : {}, {}", id, store);
>         if (store.getId() == null) {
>             throw new BadRequestAlertException("Invalid id", ENTITY_NAME, "idnull");
>         }
>         if (!Objects.equals(id, store.getId())) {
>             throw new BadRequestAlertException("Invalid ID", ENTITY_NAME, "idinvalid");
>         }
> 
>         if (!storeRepository.existsById(id)) {
>             throw new BadRequestAlertException("Entity not found", ENTITY_NAME, "idnotfound");
>         }

Reviewer 2: ```suggestion
        ...
```

        Store result = storeRepository.save(store);

        log.debug("SEND store alert for Store: {}", store);
        alertService.alertStoreStatus(result);

>         return ResponseEntity
>             .ok()
>             .headers(HeaderUtil.createEntityUpdateAlert(applicationName, true, ENTITY_NAME, store.getId()))
>             .body(result);

Reviewer 2: ```suggestion
        ...
```

            storeAlertRepository.save(storeAlert);
            emailService.sendSimpleMessage(dto);
        } catch (Exception e) {
            log.error(e.getMessage(), e);
        }
    }
}
```

**NOTE**: Any unhandled exception during message processing will make the service leave the consumer group. That's why there's code above that catches `Exception`.

As a last customization step, update the logging configuration the same way you did for the `store` microservice.

## Microservices + Kafka Container Deployment

Modify `docker-compose/docker-compose.yml` and add the following environment variables for the `alert` application:

```yaml
- SPRING_MAIL_USERNAME=${MAIL_USERNAME}
- SPRING_MAIL_PASSWORD=${MAIL_PASSWORD}
- ALERT_DISTRIBUTION_LIST=${DISTRIBUTION_LIST}
```

Edit `docker-compose/.env` and add values for the new environment variables:

```bash
MAIL_USERNAME={yourGmailAccount}
MAIL_PASSWORD={yourPassword}
DISTRIBUTION_LIST={anotherEmailAccount}
```

Make sure Docker Desktop is running, then generate the Docker image for the `store` microservice. Run the following command from the `store` directory.

```bash
./mvnw -ntp -Pprod verify jib:dockerBuild
```

> Repeat for the `alert` and `gateway` apps.

Reviewer 2: I get an error when I try running this command in the gateway app:

```
[INFO] - Generating browser application bundles (phase: setup)...
[INFO] An unhandled exception occurred: Cannot find module 'caniuse-lite/data/features/css-unicode-bidi'
[INFO] Require stack:
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/autoprefixer/data/prefixes.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/autoprefixer/lib/autoprefixer.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/postcss-preset-env/dist/index.cjs
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular-devkit/build-angular/src/webpack/configs/styles.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular-devkit/build-angular/src/webpack/configs/index.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular-devkit/build-angular/src/builders/browser/index.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular-devkit/build-angular/src/index.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular-builders/custom-webpack/dist/browser/index.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular/cli/node_modules/@angular-devkit/architect/node/node-modules-architect-host.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular/cli/node_modules/@angular-devkit/architect/node/index.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular/cli/src/command-builder/architect-base-command-module.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular/cli/src/command-builder/architect-command-module.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular/cli/src/commands/build/cli.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular/cli/src/command-builder/command-runner.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular/cli/lib/cli/index.js
[INFO] - /Users/mraible/Downloads/blog/jhipster-kafka/gateway/node_modules/@angular/cli/lib/init.js
[INFO] See "/private/var/folders/f7/5x4rvsr5599ffnfckn487nzm0000gn/T/ng-8UMHEM/angular-errors.log" for further details.
```

I'm guessing it's because I'm on airplane wi-fi right now. I'll keep trying.

Stack Overflow Question that could be related: https://stackoverflow.com/questions/73683350/version-conflict-with-angular-dependencies

Reviewer 2: It finally worked.

alert-app_1           | org.springframework.mail.MailAuthenticationException: Authentication failed; nested exception is javax.mail.AuthenticationFailedException: 535-5.7.8 Username and Password not accepted. Learn more at
alert-app_1           | 535 5.7.8  https://support.google.com/mail/?p=BadCredentials *** - gsmtp
alert-app_1           |
alert-app_1           | 	at org.springframework.mail.javamail.JavaMailSenderImpl.doSend(JavaMailSenderImpl.java:440)
```

To enable the login from the `alert` application, go to [https://myaccount.google.com](https://myaccount.google.com/) and then choose the **Security** tab. Turn on 2-Step Verification for your account. In the section _Signing in to Google_, choose **App passwords** and create a new [app password](https://support.google.com/accounts/answer/185833). In _Select app_ drop down set **Other (Custom name)** and type the name for this password. Click **Generate** and copy the password. Update `docker-compose/.env` and set the app password for gmail authentication.

```bash
MAIL_PASSWORD={yourAppPassword}
```

**IMPORTANT**: Don't forget to delete the app password once the test is done.

Restart the `alert` microservice:

```bash

> docker restart docker-compose-alert-1

Reviewer 2: This did not work for me. Whenever I changed a value in `docker-compose/.env`, I found I had to stop all the Docker containers (with Ctrl+C) and run `docker compose up` again.

Author: That's correct, updated the post

Reviewer 2: Thanks!
